### PR TITLE
Start a formal specification

### PR DESCRIPTION
**Pull request by stepchowfun**
Start a formal specification for the language. For now, the specification is just the simply-typed lambda calculus. The PDF can be built using `make spec`. [Here](https://github.com/gramlang/gram/files/863843/gram.pdf) is a pre-compiled copy.

/cc @ewang12

**Status:** Ready

**Fixes:** N/A

### CODE REVIEW

File: spec/gram.tex [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  \section*{The specification}
+
+  The specification is given incrementally. We start with the simply-typed lambda calculus. The syntax is given in Figure~\ref{fig:syntax}. Figure~\ref{fig:typing_rules} defines the typing rules. The operational semantics is listed in Figure~\ref{fig:semantics}.
+
+  \begin{figure}
+    \begin{mdframed}
+      \begin{center}
+        \begin{tabular}{l l l}
+          $\eterm \Coloneqq $ & & term \\
```
Comment (ewang12) [inline spec/gram.tex]: Why not have values be terms instead of duplicating constant and abstraction across both?

File: spec/gram.tex [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+\newcommand\hastype[2]{#1 : #2}
+\newcommand\tjudgment[3]{#1 \vdash \hastype{#2}{#3}}
+
+% Contexts
+\newcommand\ccontext{\Gamma}
+\newcommand\cempty{\varnothing}
+\newcommand\cextend[3]{#1, \ \hastype{#2}{#3}}
+
+% Operational semantics
+\newcommand\sstep[2]{#1 \longrightarrow #2}
```
Comment (ewang12) [inline spec/gram.tex]: I always thought this was a confusing arrow to use for operational semantics (easily confused with arrow types at first glance). But it's pretty standard so let's keep it.

Comment (stepchowfun) [inline spec/gram.tex]: Yeah I can see how using arrows for both makes it hard to read. Hopefully the differing arrow lengths are clear enough. I just copied the TAPL syntax.

File: spec/gram.tex (hunk 1 above)
Comment (stepchowfun) [inline spec/gram.tex]: Great idea! I'll make this change.

**Comment by stepchowfun** [comment]
Updated PDF [here](https://github.com/gramlang/gram/files/864051/gram.pdf).

**Pull request merged**